source/common/router/router.cc

```diff
@@ -1976,6 +2015,9 @@ void Filter::doRetry(bool can_send_early_data, bool can_use_http3, TimeoutRetry
   }
   UpstreamRequestPtr upstream_request = std::make_unique<UpstreamRequest>(
       *this, std::move(generic_conn_pool), can_send_early_data, can_use_http3);
+  if (retry_state_ && retry_state_->shouldRetryOnConnectionFailure() && wait_for_connect_) {
```

Comment:
this should probably be behind the runtime guard.  consider latching the runtime guard as a (temporary) member variable

Reply:
done, it was not necessary here as wait_for_connect_ would have been false when creating the 1st requests.